### ISSUE TITLE
The Case for Modularity

### ISSUE THREAD

**Issue by mvahowe** [bible-technology/scripture-burrito#88]
![big_taco](https://user-images.githubusercontent.com/1576594/68932271-8bc84900-0792-11ea-9e34-f44cf203ff53.png)
(from https://www.youtube.com/watch?v=-SnCinvntzQ)

I have been trying to think of some reasons why we keep talking past each other. I think that one factor is the difference between a typical "we just came up with an alphabet" pioneer translation project and the Bible scene for regional or world languages with large Christian communities.

The message I'm picking from those with a focus on the long tail of small languages without Scripture (ie most of the SB committee) is often "why wouldn't we keep everything in one place?" There's an argument for doing this when

- there is only one team working on this language
- most of all the work is mission funded, possibly by one organization or even one donor, so there is no commercial dimension

SB needs to support such situations. But, for larger languages, the above assumptions do not typically hold.

Even with the smallest languages, the most common scenario in DBL is that the text is owned by SIL or UBS while the audio is owned by FCBH. The text and audio tend to be cross-licensed nowadays, but the text and audio are nevertheless maintained by completely different teams with completely different tool chains and roadmaps. At the very least, the audio lags behind the latest revision for as long as it takes to make a recording. Often, the audio is many years and revisions behind the text. If the audio is revised, it may skip many text revisions. In some cases it may purposely not use the most recent text revision (for example if the audio it to be used with a text that has already been published.)

Audio Bibles used to be a niche market but, today, they are seen as a potential revenue stream, and the availablility of good, cheap recording technology has encouraged many new initiatives. For one French translation I am aware of recordings with three different accents. Dramatized recordings are popular right now. Each of these recordings may be made by a different organization.

From a technical perspective, keeping everything in one burrito is appealing for the same reason that making all variables global is really cool. Everything is easy to do, the system never tells you "no"... but we've all done this and all learned, the hard way, that sometimes it's better to put Chinese walls between things.

For languages like French, keeping everything related to a project in one burrito tends to look like the photo above (which, incidentally, is a 4lb taco). In some cases the project might enter the Gb range.

Also,  the content of that megaburrito may be owned by many different organizations. I know of one case where the OT and NT recordings are owned by different organizations. Putting everything in one place would create the mother of all VC headaches, across all those organizations. It would also mean that, eg, a Paratext user working on a new text revision would need to care about a gigabyte of audio from the previous revision.

Once the new text is ready, as a new revision, what happens to the audio that will never match that new revision, and may not match the previous revision either, but which is stored in the same burrito? Do I need the permission of 10 organizations to distribute this megaburrito or, alternatively, can any one organization decide for the others? And so on.

There's a cost to modularity. Those working on a project will sometimes need to work with several related "things" rather than one "thing" that has everything they need. Also, links between those "things" typically need to be coarser than in the "one thing" scenario, for the same reason that classes are not supposed to rely on the internals of other classes. Both these constraints can be frustrating. But the alternative is simply unworkable as soon as we get beyond the one-team Bible ecosystem.

This is why I favour breaking content down in to the smallest units that make sense. "Making sense" is of course a judgment call, and that' the job of this committee. But I think it's simply a fact that the units need to be smaller than "everything related to a translaton project".

**Comment by jag3773** [comment]
I agree that smaller burritos with more linking capabilities is the way to go.  I think that many of the use cases have pretty clear dividing lines that aren't too hard to identify, but you are right that there are several which will be tricky to identify.

For example, mixed-media projects strike me as a difficult situation. For example, our Open Bible Stories are text and images, they belong together, but we could make the argument that the images should reside in their own burrito. One advantage of splitting them is that the images are the same for all translations (or most), so any number of text burritos could link to the same image burrito (which would cut down on app size for example). The downside is that we create a scenario where the text and images can be divorced and hence, likely will be in certain cases, even if that's not intended.

One problem we've encountered using [Resource Container links](https://resource-container.readthedocs.io/en/latest/linking.html) is that it is difficult to make them work in all environments. For example, on the web you really just want a URL which is understood by all web technologies–It's painful to have to create web based software to utilize custom links. However, in an app it doesn't matter so much because you've got to code something anyway, but if you want your app to work offline then you don't really want it to have to load URLs.

At the moment I'm blurring the lines a bit between linking in our metadata between burritos and also linking to other content within the content of a burrito. I think we are going to need to address both things in a robust manner.

**Comment by jonathanrobie** [comment]
> Even with the smallest languages, the most common scenario in DBL is that
the text is owned by SIL or UBS while the audio is owned by FCBH.

I think that's the status quo for most projects.  In translations where SIL
or UBS owns the text and FCBH owns the audio, there is clearly more than
one project.

But increasingly, Paratext projects are also using oral translation
techniques and FCBH projects also need text.  In each case, whatever the
project participants think of as "one project" might best be put in one SB
project.   In other words, I'm not convinced that the media type should
determine what a project is.

**Comment by mvahowe** [comment]
> In each case, whatever the project participants think of as "one project" might best be put in one SB project.

I don't think that's workable. SB is at the data level, and I think we need to make decisions about data models on the basis of what makes engineering sense. I don't think that building systems to work according to end user models, at a low level, is the way to go. Apart from anything else, I'm pretty sure that 3 users will give you 4 different answers.

Making the underlying data model make sense to users via the UX is an important problem, but it's a different problem. So, eg, a processor could automagically grab related audio when a text project is loaded, or vice versa, and keep the fact that there are 2 burritos involved a secret. (Of course this assumes a server ecosystem :-) ) Conceptually, this is no different to the way most IT systems present a single UX table per page, but that table may be the result of joining multiple tables under the hood.
To put it differently, you can think of a mesh of related burritos like a distributed document database.

> In other words, I'm not convinced that the media type should determine what a project is.

"Media type" is a really unhelpful word here, and is not what distinguishes flavors, at all. The key differences between scriptureText and scriptureAudio are not primarily the mimetypes of the resources. Audio projects are structured differently, have different metadata, and so on. scriptureText, scripturePrint and scriptureBraille are all "text" but do very different things. We might need both text and audio, or both text and video, or something else, in one flavor, but I don't think we can always have "everything that someone thinks is a project" in one flavor.

**Comment by mvahowe** [comment]
I feel like we all said what we needed to say and that the conversation has moved on, so can we close this?
